Source/Core/AudioCommon/Mixer.cpp

```diff
@@ -49,128 +74,154 @@ unsigned int CMixer::MixerFifo::Mix(short* samples, unsigned int numSamples, boo
+	// for each output sample pair (left and right),
+	// linear interpolate between current and next sample
+	// increment output sample position
+	// increment input sample position by ratio, store fraction
+	// QUESTION: do we need to check for NUM_CROSSINGS samples before we interpolate?
+	// seems to work fine as is
+	for (; current_sample < numSamples * 2 && ((write_index - read_index) & INDEX_MASK) > 0; current_sample += 2)
 	{
-		u32 indexR2 = indexR + 2; //next sample
-
-		s16 l1 = Common::swap16(m_buffer[indexR & INDEX_MASK]); //current
-		s16 l2 = Common::swap16(m_buffer[indexR2 & INDEX_MASK]); //next
-		int sampleL = ((l1 << 16) + (l2 - l1) * (u16)m_frac) >> 16;
-		sampleL = (sampleL * lvolume) >> 8;
-		sampleL += samples[currentSample + 1];
-		samples[currentSample + 1] = MathUtil::Clamp(sampleL, -32767, 32767);
-
-		s16 r1 = Common::swap16(m_buffer[(indexR + 1) & INDEX_MASK]); //current
-		s16 r2 = Common::swap16(m_buffer[(indexR2 + 1) & INDEX_MASK]); //next
-		int sampleR = ((r1 << 16) + (r2 - r1) * (u16)m_frac) >> 16;
-		sampleR = (sampleR * rvolume) >> 8;
-		sampleR += samples[currentSample];
-		samples[currentSample] = MathUtil::Clamp(sampleR, -32767, 32767);
-
-		m_frac += ratio;
-		indexR += 2 * (u16)(m_frac >> 16);
-		m_frac &= 0xffff;
+		float l_output, r_output;
+		Interpolate(read_index, &l_output, &r_output);
+		samples[current_sample + 1] += l_volume * l_output;
+		samples[current_sample] += r_volume * r_output;
+		m_fraction += ratio;
+		read_index += 2 * (s32)m_fraction;
+		m_fraction = m_fraction - (s32)m_fraction;
 	}
-
-	// Padding
-	short s[2];
-	s[0] = Common::swap16(m_buffer[(indexR - 1) & INDEX_MASK]);
-	s[1] = Common::swap16(m_buffer[(indexR - 2) & INDEX_MASK]);
-	s[0] = (s[0] * rvolume) >> 8;
-	s[1] = (s[1] * lvolume) >> 8;
-	for (; currentSample < numSamples * 2; currentSample += 2)
+	// pad output if not enough input samples
+	float s[2];
+	s[0] = m_float_buffer[(read_index - 1) & INDEX_MASK] * r_volume;
+	s[1] = m_float_buffer[(read_index - 2) & INDEX_MASK] * l_volume;
+	for (; current_sample < numSamples * 2; current_sample += 2)
 	{
-		int sampleR = MathUtil::Clamp(s[0] + samples[currentSample + 0], -32767, 32767);
-		int sampleL = MathUtil::Clamp(s[1] + samples[currentSample + 1], -32767, 32767);
-
-		samples[currentSample + 0] = sampleR;
-		samples[currentSample + 1] = sampleL;
+		samples[current_sample] += s[0];
+		samples[current_sample + 1] += s[1];
 	}
+	// update read index
+	m_read_index.store(read_index);
+}
 
-	// Flush cached variable
-	m_indexR.store(indexR);
+u32 CMixer::MixerFifo::AvailableSamples()
+{
+	return ((m_write_index.load() - m_read_index.load()) & INDEX_MASK) * 48000 / (2 * m_input_sample_rate);
+}
 
-	return numSamples;
+u32 CMixer::AvailableSamples()
+{
+	u32 samples = m_dma_mixer.AvailableSamples();
+	if (samples == 0)
+	{
+		samples = m_streaming_mixer.AvailableSamples();
+	}
+	if (samples == 0)
+	{
+		samples = m_wiimote_speaker_mixer.AvailableSamples();
```

Comment:
This function ignores that all of those may have different sample rates. So in the end, this function likely isn't useful.